src/test/java/com/aws/greengrass/componentmanager/KernelConfigResolverTest.java

```diff
@@ -386,6 +386,213 @@ void GIVEN_deployment_with_artifact_WHEN_config_resolution_requested_THEN_artifa
+        List<ComponentIdentifier> packagesToDeploy = Arrays.asList(rootComponentIdentifier);
+
+        ComponentRecipe rootComponentRecipe = new ComponentRecipe(RecipeFormatVersion.JAN_25_2020, TEST_INPUT_PACKAGE_A,
+                rootComponentIdentifier.getVersion(), "", "", null, new HashMap<String, Object>() {{
+            put(LIFECYCLE_RUN_KEY, "java -jar {configuration:/jarPath}/test.jar -x arg");
+        }}, Collections.emptyList(), Collections.emptyMap(), null);
+
+        DeploymentPackageConfiguration rootPackageDeploymentConfig = DeploymentPackageConfiguration.builder()
+                .packageName(TEST_INPUT_PACKAGE_A)
+                .rootComponent(true)
+                .resolvedVersion("=1.2")
+                .configurationUpdateOperation(new ConfigurationUpdateOperation(
+                        new HashMap<String, Object>() {{
+                            put("jarPath", "{artifacts:path}");
+                        }}, new ArrayList<String>()))
+	        .build();
+        DeploymentDocument document = DeploymentDocument.builder()
+                                                        .deploymentPackageConfigurationList(
+                                                                Arrays.asList(rootPackageDeploymentConfig))
+                                                        .timestamp(10_000L)
+                                                        .build();
+
+        when(componentStore.getPackageRecipe(rootComponentIdentifier)).thenReturn(rootComponentRecipe);
+        when(nucleusPaths.artifactPath(rootComponentIdentifier)).thenReturn(Paths.get("/packages/artifacts"));
+        when(kernel.getMain()).thenReturn(mainService);
+        when(mainService.getName()).thenReturn("main");
+        when(mainService.getDependencies()).thenReturn(Collections.emptyMap());
+        when(deviceConfiguration.getInterpolateConfiguration()).thenReturn(config.lookup("interpolateConfiguration").withValue(true));
+
+        // WHEN
+        KernelConfigResolver kernelConfigResolver = new KernelConfigResolver(componentStore, kernel, nucleusPaths,
+                deviceConfiguration);
+        Map<String, Object> resolvedConfig =
+                kernelConfigResolver.resolve(packagesToDeploy, document, Arrays.asList(TEST_INPUT_PACKAGE_A));
+
+        // THEN
+        Map<String, Object> servicesConfig = (Map<String, Object>) resolvedConfig.get(SERVICES_NAMESPACE_TOPIC);
+
+        Path jarPath = Paths.get("/packages/artifacts").toAbsolutePath();
+        assertThat("{configuration:/jarPath} should be replace by the package's artifact path",
+                getServiceRunCommand(TEST_INPUT_PACKAGE_A, servicesConfig),
+                equalTo("java -jar " + jarPath + "/test.jar -x arg"));
+    }
+
+    @Test
+    void GIVEN_interpolate_config_set_WHEN_config_resolution_requested_THEN_self_referential_config_resolved_one_level()
```

Comment:
Do you think this is the appropriate/best/most understandable behavior to only resolve a single level instead of trying to recursively resolve until the interpolation value is found or cannot be resolved?

Reply:
That's a good question - I would suspect for most usescases resolving a single level like this is enough. I had started with this way because the alternative gets a bit more complicated, not just with dealing with figuring out when to stop resolving, but also either detecting loops or disallowing a component from referencing it's own config (which would mean needing slightly different codepaths between interpolation for lifecycle vs config).

With the behavior as implemented you can still sort of get two levels of resolution if the first is in the config and the second is in the lifecycle, and at least for usecases I have that's enough - e.g. I can have `{iot:ThingName}` in the config for an environment variable in a lambda component, and that component interpolates `{configuration:/lambdaExecutionParameters}` (where those env vars are set) into it's setenv, and because configuration is interpolated first, and then lifecycle is interpolated after that first interpolation is applied, that works.

Reply:
Yes that sounds good to me. If we have a requirement later on it sounds like we could add that ability without issue (not a one-way door).